rpc/serve.go

```diff
@@ -0,0 +1,72 @@
+package rpc
+
+import (
+	"context"
+	"errors"
+	"net"
+
+	"capnproto.org/go/capnp/v3"
+)
+
+// Serve serves a Cap'n Proto RPC to incoming connections.
+//
+// Serve will take ownership of bootstrapClient and release it after the listener closes.
+//
+// Serve exits with the listener error if the listener is closed by the owner.
+func Serve(lis net.Listener, bootstrapClient capnp.Client) error {
+	if !bootstrapClient.IsValid() {
+		err := errors.New("BootstrapClient is not valid")
+		return err
+	}
+	// Accept incoming connections
+	for {
+		rwc, err := lis.Accept()
+		if err != nil {
+			// Since we took ownership of the bootstrap client, release it after we're done.
+			if !bootstrapClient.IsValid() {
+				err = errors.New("the bootstrap client was already released")
+			}
+			bootstrapClient.Release()
+			return err
+		}
+
+		// For each new incoming connection, create a new RPC transport connection that will serve incoming RPC requests
+		// rpc.Options will contain the bootstrap capability
+		go func() {
+			// skip if the bootstrap client has closed since receiving the connection
+			// this can happen if the server exits while incoming connections are made
+			if bootstrapClient.IsValid() {
+				transport := NewStreamTransport(rwc)
+				// the RPC connection takes ownership of the bootstrap interface and will release it when the connection
+				// exits, so use AddRef to avoid releasing the provided bootstrap client capability.
+				opts := Options{
+					BootstrapClient: bootstrapClient.AddRef(),
```

Comment:
We should hoist the call to .AddRef() up above, before we fork off a goroutine. This avoids a scenario where:

- We fork off a goroutine to handle one connection
- The next call to Accept() fails for some reason, so we return, dropping the reference to bootstrapClient
- The goroutine we forked tries to .AddRef(), and it panics since bootstrapClient has already been released.

Reply:
Well spotted. This is exactly the problem I ran into and described above. Excellent point and great solution!